generate-osm/db-utils/find_coordinates.go

```diff
@@ -0,0 +1,317 @@
+			runningWay = startWay[0]
+			downId, downDist = goDown(runningWay, 0, dist)
+		} else if startWay[1].Nd[0].Ref == node.Id && startWay[0].Nd[len(startWay[0].Nd)-1].Ref == node.Id {
+			runningWay := startWay[0]
+			upId, upDist = goUp(runningWay, len(runningWay.Nd)-1, dist)
+			runningWay = startWay[1]
+			downId, downDist = goDown(runningWay, 0, dist)
+		} else {
+			err = errors.New("Error with ways!")
+			return
+		}
+	default: 
+		err = errors.New("Too many ways!")
+		return
+	}
+	err = nil
+	return 
+}
+
+func goUp(runningWay OSMUtil.Way, index int, dist float64) (string, float64) {	
+	runningNode, _ := getNode(runningWay.Nd[index].Ref)
+	var oldNode *OSMUtil.Node
+	var nextNode *OSMUtil.Node
+	totalDist := 0.0
+	wayDirUp := true
+
+	for ; totalDist < dist; {
+		runningWay, index, wayDirUp, nextNode = findNextWay(wayDirUp, index, runningNode, oldNode, runningWay)
+
+		if index == -1 {
+			return runningNode.Id, totalDist
+		}
+		
+		phi1, _ := strconv.ParseFloat(runningNode.Lat, 64)
+		phi2, _ := strconv.ParseFloat(nextNode.Lat, 64)
+		lambda1, _ := strconv.ParseFloat(runningNode.Lon, 64)
+		lambda2, _ := strconv.ParseFloat(nextNode.Lon, 64)
+		
+		totalDist += distance(phi1, phi2, lambda1, lambda2)
+
+		if totalDist == dist {
+			print("Dist is gud \n")
+			return runningWay.Nd[index].Ref, totalDist		
+		}
+
+		if wayDirUp {
+			index--
+		} else {
+			index++
+		}
+		oldNode = runningNode
+		runningNode = nextNode
+	}
+	return runningNode.Id, totalDist
+}
+
+func goDown(runningWay OSMUtil.Way, index int, dist float64) (string, float64) {	
+	runningNode, _ := getNode(runningWay.Nd[index].Ref)	
+	var oldNode *OSMUtil.Node
+	var nextNode *OSMUtil.Node
+	totalDist := 0.0
+	wayDirUp := false
+
+	for ; totalDist < dist; {
+		runningWay, index, wayDirUp, nextNode = findNextWay(wayDirUp, index, runningNode, oldNode, runningWay)
+
+		if index == -1 {
+			return runningNode.Id, totalDist
+		}
+
+		phi1, _ := strconv.ParseFloat(runningNode.Lat, 64)
+		phi2, _ := strconv.ParseFloat(nextNode.Lat, 64)
+		lambda1, _ := strconv.ParseFloat(runningNode.Lon, 64)
+		lambda2, _ := strconv.ParseFloat(nextNode.Lon, 64)
+
+		totalDist += distance(phi1, phi2, lambda1, lambda2)		
+
+		if totalDist == dist {
+			print("Dist is gud \n")
+			return runningWay.Nd[index].Ref, totalDist		
+		}
+
+		if wayDirUp {
+			index--
+		} else {
+			index++
+		}
+		oldNode = runningNode
+		runningNode = nextNode
+	}
+	return runningNode.Id, totalDist
+}
+
+func findNextWay(wayDirUp bool, index int, runningNode *OSMUtil.Node, oldNode *OSMUtil.Node, runningWay OSMUtil.Way) (OSMUtil.Way, int, bool, *OSMUtil.Node) {
+	if wayDirUp && index == 0 {
+		nextWays, err := findWay(runningNode.Id)
+		if err != nil || len(nextWays) == 0 {
+			panic(errors.New("No ways!"))
+		}
+		if len(nextWays) == 1 || len(nextWays) > 2 {
+			return runningWay, -1, false, nil
+		}
+
+		if nextWays[0].Nd[0].Ref == nextWays[1].Nd[len(nextWays[1].Nd)-1].Ref && nextWays[0].Nd[0].Ref == runningNode.Id {

```

Comment:
Too complicated condition. Extract into bool before. And give it a good name so the if is more readable.
Or mabye even rewrite this. This hugh if is really hard to understand